### ISSUE TITLE
Convert all the existing `examples` arrays on service classes to `openApi` objects

### ISSUE THREAD

**Issue by chris48s** [badges/shields#9285]
:clipboard: **Description**

Now that we have launched the new frontend, the current situation we are in is:

- Service classes may either define an `openApi` spec directly, or a legacy `examples` array
- There are a small handful of service classes which define an `openApi` spec directly, for example https://github.com/badges/shields/blob/master/services/dynamic/dynamic-xml.service.js#L18
- Most services still define a legacy `examples` array
- For services defining a legacy `examples` array, we make some assumptions and [bash it](https://github.com/badges/shields/blob/30377f6f33f7506364911aa034bb6d9ad188111a/core/base-service/openapi.js) into an Open API spec at frontend build time as best we can (accepting there are some limitations)

We want to move to a situation where every service class defines an `openApi` object, so step one is to auto-convert the legacy `examples` array to an `openApi` object as best we can. Then we can start tweaking/improving from there.

This is going to be a big PR touching basically every .service.js file in the codebase.

**Comment by chris48s** [comment]
I'm not going to raise individual issues for these tasks yet as they are all blocked on this, but the next steps after this will be:

- Delete all `examples` properties + remove any validation for it
- Make `openApi` property required/validated
- Delete all the code for bludgeoning legacy `examples` objects to OpenAPI objects
- Correctly document required query params
- Fix duplicated titles
- Remove unnecessary duplicates for query param variants
- Start adding param descriptions
- Update contributing docs for adding new services to reflect new frontend + documenting with `openApi` property instead of `examples`
- Look for ways we can reduce duplication between the service definition/schema and OpenAPI spec
- Implement some automated checks (be that Joi schema, CI checks - TBC) that help to prevent common issues

Some of those might end up getting done in the first PR. I'll split the rest out into issues once they are unblocked.

**Comment by calebcartwright** [comment]
I want to be clear up front that I'm **not** arguing _against_ doing this. However, I would like to know what, if any, level of coupling risk exists here.

I.e. if we convert everything over to openapi and then decide we need to e.g. go back to old UI or something different, would we be able to easily continue using the existing openapi-based example listings defined in each service class, and/or centrally "convert" them at the frontend build time like we're doing today (a openapi -> foo translation vs. the current legacyExamples -> openapi)

**Comment by chris48s** [comment]
In general, one of the advantages of defining the shields URL schema using OpenAPI is that it is a widely used specification and many tools can consume/render it. So we're way less coupled in that respect. If we decide that docusaurus-openapi is terrible in future then we'd be more likely to be able to migrate to some other frontend with minimal changes to the service classes.

Conversely the old `examples` array was a format we made. It was completely bespoke to the old frontend and hence there is/was a one-to-one coupling.

If we do think there is serious risk that we want to go back to the old frontend at this point, I'd rather put things on hold and work out what we need to do to alleviate the concerns before moving on, rather than try to press on but build in a backwards migration path.

**Comment by calebcartwright** [comment]
That makes sense to me, ty. My impression was that this is/would continue to be a similar paradigm to other parts of the base class hierarchy, e.g. request parameters and maintaining the ability to do any necessary conversion/translation to what a 3rd party lib we're using might need

**Comment by chris48s** [comment]
I've been having a bit of a think about how to do the bulk conversion. It basically requires a kind of "syntax aware find and replace".

I've been having a think about options. I have not tried it out yet, but https://github.com/facebook/jscodeshift looks very promising. I think I am going to look into this first.

**Comment by chris48s** [comment]
@PyvesB has really helped get through this backlog of reviews over the last few days.

I am going to see if I can take another chunk out of this job in the gap between Christmas and New Year.

**Comment by chris48s** [comment]
I think we now have at least an open PR covering everything we need to migrate.

**Comment by chris48s** [comment]
Right. Merged all the PRs and it turns out.. missed a spot :facepalm: There is one more left :laughing: 

Some time soon - probably at the weekend - I will do that final one and also look at the next job, which is removing all the messy compatibility layer that allows both `examples[]` and `openApi{}` to exist side by side and make the `openApi` property required/validated. There is a lot of annoying code I'll be able to delete :)
